modules/standard/Memory/Initialization.chpl

```diff
@@ -0,0 +1,290 @@
+/*
+ * Copyright 2020-2021 Hewlett Packard Enterprise Development LP
+ * Copyright 2004-2019 Cray Inc.
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+  The :mod:`Initialization` module provides functions which enable users to
+  move-initialize and deinitialize values.
+
+  The functionality provided by this module can be used to implement
+  collections in a manner similar to those implemented by the Chapel standard
+  modules (such as :mod:`List` or :mod:`Map`).
+*/
+module Initialization {
+
+  private inline proc _move(ref dst: ?t, const ref src: t) {
+    __primitive("=", dst, src);
+  }
+
+  private inline proc _destroy(ref r) {
+    chpl__autoDestroy(r);
+  }
+
+  private inline proc _needsAutoDestroy(type t) param: bool {
+    return __primitive("needs auto destroy", t);
+  }
+
+  /*
+    Check to see if a given type supports deinitialization.
+
+    :arg t: A type to check for a deinitializer
+    :type t: `type`
+
+    :return: ``true`` if ``t`` supports deinitialization
+    :rtype: param bool
+  */
+  proc needsDeinit(type t) param {
+    return _needsAutoDestroy(t);
+  }
+
+  /*
+    Explicitly deinitialize a reference. The memory referred to by ``arg``
+    should be considered uninitialized after a call to this function.
+
+    .. warning::
+
+      Chapel is not aware of a call to ``explicitDeinit()``, which means
+      that the deinitializer of ``arg`` may run again once it goes out of
+      scope.
+
+    :arg: A value to deinitialize
+  */
+  proc explicitDeinit(ref arg: ?t) {
+    if needsDeinit(t) then
+      _destroy(arg);
+  }
+
+  /*
+    Move-initialize a reference from a value. No assignment is performed
+    when moving ``rhs`` into the memory referred to by ``lhs``.
```

Comment:
This function can run the copy initializer today, for example if the actual argument for `rhs` is a reference. We should make that case into a compilation error but it is OK to leave that as future work.

Reply:
(Just be sure to make a TODO issue that we need to do that follow up task).

Reply:
>This function can run the copy initializer today

Uh-huh, then it is not a `move` and we need to do more explaining.

Maybe present it as equivalent to `proc init=(in rhs)` except (a) it can be invoked explicitly and so (b) overwrites whatever was in `lhs` before ?

Reply:
Oh, I don't think we should present it that way. I view it as a bug that it can run the copy initializer today. We could put a `.. note::` section about it if you want, but I hope David just makes a future work issue and we attend to it soon.

Reply:
I am hesitant to use the word `ownership` here. To me it implies `owned` classes. How about something like...

```
    Move-initialize ``lhs`` with the value in ``rhs``. The contents of ``rhs``
    are not deinitialized before the move, and ``rhs`` is not deinitialized
    after the move.
```